crypto/src/lib.rs

```diff
@@ -547,37 +456,152 @@ impl PrivateKey {
+    impl std::error::Error for MultihashConvertError {}
+
+    impl From<NoSuchAlgorithm> for MultihashConvertError {
+        fn from(_: NoSuchAlgorithm) -> Self {
+            Self {
+                reason: String::from("Digest function not supported"),
+            }
+        }
+    }
+
+    /// Error which occurs when parsing [`PublicKey`]
+    #[derive(Debug, Display, Clone)]
+    pub enum KeyParseError {
+        /// Decoding hex failed
+        #[display(fmt = "Failed to decode. {_0}: {_1}")]
+        Decode(String, hex::FromHexError),
+        /// Converting bytes to multihash failed
+        #[display(fmt = "Failed to convert. {_0}: {_1}")]
+        Multihash(String, MultihashConvertError),
+    }
+
+    #[cfg(feature = "std")]
+    impl std::error::Error for KeyParseError {}
+    /// Error when dealing with cryptographic functions
+    #[derive(Debug, Display)]
+    pub enum Error {
+        /// Returned when trying to create an algorithm which does not exist
+        #[display(fmt = "Algorithm doesn't exist")] // TODO: which algorithm
+        NoSuchAlgorithm,
+        /// Occurs during deserialization of a private or public key
+        #[display(fmt = "Key could not be parsed. {_0}")]
+        Parse(String),
+        /// Returned when an error occurs during the signing process
+        #[display(fmt = "Signing failed. {_0}")]
+        Signing(String),
+        /// Returned when an error occurs during key generation
+        #[display(fmt = "Key generation failed. {_0}")]
+        KeyGen(String),
+        /// Returned when an error occurs during digest generation
+        #[display(fmt = "Digest generation failed. {_0}")]
+        DigestGen(String),
+        /// Returned when an error occurs during creation of [`SignaturesOf`]
+        #[display(fmt = "`SignaturesOf` must contain at least one signature")]
+        EmptySignatureIter,
+        /// A General purpose error message that doesn't fit in any category
+        #[display(fmt = "General error. {_0}")] // This is going to cause a headache
+        Other(String),
+    }
+
+    #[cfg(feature = "std")]
+    #[cfg(not(feature = "ffi_import"))]
+    impl From<ursa::CryptoError> for Error {
+        fn from(source: ursa::CryptoError) -> Self {
+            match source {
+                ursa::CryptoError::NoSuchAlgorithm(_) => Self::NoSuchAlgorithm,
+                ursa::CryptoError::ParseError(source) => Self::Parse(source),
+                ursa::CryptoError::SigningError(source) => Self::Signing(source),
+                ursa::CryptoError::KeyGenError(source) => Self::KeyGen(source),
+                ursa::CryptoError::DigestGenError(source) => Self::DigestGen(source),
+                ursa::CryptoError::GeneralError(source) => Self::Other(source),
+            }
+        }
+    }
+
+    #[cfg(feature = "std")]
+    impl From<NoSuchAlgorithm> for Error {
+        fn from(_: NoSuchAlgorithm) -> Self {
+            Self::NoSuchAlgorithm
+        }
+    }
```

Comment:
This should be implemented regardless of `std`. What should be made conditional is setting the `source` of the error (which doesn't take place here).